flambe/nn/transformer_sru.py

```diff
@@ -0,0 +1,307 @@
+        """Initialize a TransformerDecoder.
+
+        Parameters
+        ----------
+        d_model : int
+            The number of expected features in the input.
+        n_head : int
+            The number of heads in the multiheadattention models.
+        dim_feedforward : int, optional
+            The dimension of the feedforward network (default=2048).
+        dropout : float, optional
+            The dropout value (default=0.1).
+
+        Extra keyword arguments are passed to the SRUCell.
+
+        """
+        super(TransformerSRUDecoderLayer, self).__init__()
+
+        self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.multihead_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.sru = SRUCell(d_model,
+                           dim_feedforward,
+                           dropout,
+                           bidirectional=False,
+                           has_skip_term=False, **kwargs)
+
+        self.linear2 = nn.Linear(dim_feedforward, d_model)
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.norm3 = nn.LayerNorm(d_model)
+        self.dropout1 = nn.Dropout(dropout)
+        self.dropout2 = nn.Dropout(dropout)
+        self.dropout3 = nn.Dropout(dropout)
+
+    def forward(self,  # type: ignore
+                tgt: torch.Tensor,
+                memory: torch.Tensor,
+                tgt_mask: Optional[torch.Tensor] = None,
+                memory_mask: Optional[torch.Tensor] = None,
+                tgt_key_padding_mask: Optional[torch.Tensor] = None,
+                memory_key_padding_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        r"""Pass the inputs (and mask) through the decoder layer.
+
+        Parameters
+        ----------
+        tgt: torch.Tensor
+            The sequence to the decoder layer (required).
+        memory: torch.Tensor
+            The sequnce from the last layer of the encoder (required).
+        tgt_mask: torch.Tensor, optional
+            The mask for the tgt sequence (optional).
+        memory_mask: torch.Tensor, optional
+            the mask for the memory sequence (optional).
+        tgt_key_padding_mask: torch.Tensor, optional
+            the mask for the tgt keys per batch (optional).
+        memory_key_padding_mask: torch.Tensor, optional
+            the mask for the memory keys per batch (optional).
+
+        """
+        tgt2 = self.self_attn(tgt, tgt, tgt, attn_mask=tgt_mask,
+                              key_padding_mask=tgt_key_padding_mask)[0]
+        tgt = tgt + self.dropout1(tgt2)
+        tgt = self.norm1(tgt)
+        tgt2 = self.multihead_attn(tgt, memory, memory, attn_mask=memory_mask,
+                                   key_padding_mask=memory_key_padding_mask)[0]
+        tgt = tgt + self.dropout2(tgt2)
+        tgt = self.norm2(tgt)
+        tgt2 = self.linear2(self.sru(tgt, state=memory, mask_pad=tgt_key_padding_mask))
```

Comment:
`state` is not an argument for SRUcell. (should have renamed `c0` into `state` in the future for SRU)
for now you can do `self.sru(tgt, memory, mask_pad=..)` instead